File: backend/app/api/papers.py
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from typing import List, Optional
from app.database.database import get_db
from app.models.models import Paper, Subject, Class
from app.schemas.schemas import PaperOut, PaperDetail, SearchResult, SearchResponse

router = APIRouter(tags=["Papers"])

EXAM_TYPES = [
    "Unit Test 1", "Unit Test 2", "Unit Test 3",
    "Quarterly Exam", "Half Yearly Exam",
    "Annual Exam", "Public Exam",
    "Practical Exam", "Model Exam",
]


@router.get("/exam-types")
def get_exam_types():
    return {"exam_types": EXAM_TYPES}


@router.get("/subjects/{subject_id}/papers", response_model=List[PaperOut])
def get_papers_for_subject(
    subject_id: int,
    exam_type: Optional[str] = None,
    paper_type: Optional[str] = None,
    year: Optional[int] = None,
    db: Session = Depends(get_db),
):
    subject = db.query(Subject).filter(Subject.id == subject_id).first()
    if not subject:
        raise HTTPException(status_code=404, detail="Subject not found")

    q = db.query(Paper).filter(Paper.subject_id == subject_id, Paper.is_visible == True)
    if exam_type:
        q = q.filter(Paper.exam_type == exam_type)
    if paper_type:
        q = q.filter(Paper.paper_type == paper_type)
    if year:
        q = q.filter(Paper.year == year)

    return q.order_by(Paper.year.desc()).all()


@router.get("/papers/{paper_id}", response_model=PaperDetail)
def get_paper(paper_id: int, db: Session = Depends(get_db)):
    paper = db.query(Paper).filter(Paper.id == paper_id, Paper.is_visible == True).first()
    if not paper:
        raise HTTPException(status_code=404, detail="Paper not found")
    return paper


@router.get("/search", response_model=SearchResponse)
def search_papers(
    q: str = Query(..., min_length=1, description="Search query"),
    class_id: Optional[int] = None,
    exam_type: Optional[str] = None,
    paper_type: Optional[str] = None,
    year: Optional[int] = None,
    db: Session = Depends(get_db),
):
    query = (
        db.query(Paper)
        .join(Subject, Paper.subject_id == Subject.id)
        .join(Class, Subject.class_id == Class.id)
        .filter(Paper.is_visible == True)
    )

    search_term = f"%{q.lower()}%"
    from sqlalchemy import or_, func
    query = query.filter(
        or_(
            func.lower(Paper.title).like(search_term),
            func.lower(Paper.exam_type).like(search_term),
            func.lower(Subject.name).like(search_term),
        )
    )

    if class_id:
        query = query.filter(Subject.class_id == class_id)
    if exam_type:
        query = query.filter(Paper.exam_type == exam_type)
    if paper_type:
        query = query.filter(Paper.paper_type == paper_type)
    if year:
        query = query.filter(Paper.year == year)

    papers = query.order_by(Paper.created_at.desc()).limit(50).all()

    results = [
        SearchResult(
            id=p.id,
            title=p.title,
            exam_type=p.exam_type,
            year=p.year,
            paper_type=p.paper_type,
            subject_name=p.subject.name,
            class_name=p.subject.class_.name,
            public_url=p.public_url,
        )
        for p in papers
    ]

    return SearchResponse(query=q, total=len(results), results=results)

File: frontend/src/services/papers.js
// =============================================================================
// Papers Service — migrated to FastAPI
// =============================================================================
// All data-fetching requests now go through FastAPI DEV:
//   GET /api/v1/papers?sort=recent&limit=N   ← getRecentPapers
//   GET /api/v1/papers?sort=popular&limit=N  ← getPopularPapers
//   GET /api/v1/papers/{paper_id}            ← getPaper
//   POST /api/v1/papers/{paper_id}/download  ← recordDownload
//
// Public function signatures and return shapes are preserved so that
// all components (HomePage, PaperDetailPage) require zero changes.
//
// NOTE: Constants (EXAM_TYPES, MONTHS, TN_DISTRICTS) are pure data —
// no Supabase involved — they remain unchanged.
// =============================================================================

import { apiFetch } from '../lib/api'

// =============================================================================
// Paper metadata constants — single source of truth for the entire app.
// Import these in admin pages and search — do NOT duplicate locally.
// =============================================================================

export const EXAM_TYPES = [
  'Monthly Test',
  'First Mid Term Test',
  'Unit Test 1',
  'Unit Test 2',
  'Unit Test 3',
  'Quarterly Exam',
  'Half Yearly Exam',
  'Annual Exam',
  'Public Exam',
  'Practical Exam',
  'Model Exam',
]

export const MONTHS = [
  'January', 'February', 'March', 'April',
  'May', 'June', 'July', 'August',
  'September', 'October', 'November', 'December',
]

// All 38 Tamil Nadu districts — alphabetically sorted.
export const TN_DISTRICTS = [
  'Ariyalur',
  'Chengalpattu',
  'Chennai',
  'Coimbatore',
  'Cuddalore',
  'Dharmapuri',
  'Dindigul',
  'Erode',
  'Kallakurichi',
  'Kancheepuram',
  'Kanyakumari',
  'Karur',
  'Krishnagiri',
  'Madurai',
  'Mayiladuthurai',
  'Nagapattinam',
  'Namakkal',
  'Nilgiris',
  'Perambalur',
  'Pudukkottai',
  'Ramanathapuram',
  'Ranipet',
  'Salem',
  'Sivaganga',
  'Tenkasi',
  'Thanjavur',
  'Theni',
  'Thoothukudi',
  'Tiruchirappalli',
  'Tirunelveli',
  'Tirupathur',
  'Tiruppur',
  'Tiruvallur',
  'Tiruvannamalai',
  'Tiruvarur',
  'Vellore',
  'Villupuram',
  'Virudhunagar',
]

// =============================================================================
// Adapter — FastAPI flat PaperResponse → Supabase-compatible nested shape
// =============================================================================
// FastAPI GET /api/v1/papers/{id} returns a flat object:
//   { id, subject_id, subject_name, subject_slug, class_id, class_name, ... }
//
// PaperDetailPage expects the old Supabase nested shape:
//   paper.subjects.name
//   paper.subjects.classes.name
//   paper.subjects.class_id
//
// This adapter reconstructs the nested shape from the flat fields so that
// PaperDetailPage requires zero changes.
// =============================================================================

function adaptPaperDetail(p) {
  return {
    ...p,
    subjects: {
      id:           p.subject_id,
      name:         p.subject_name   ?? null,
      slug:         p.subject_slug   ?? null,
      is_practical: p.is_practical   ?? null,
      class_id:     p.class_id       ?? null,
      classes: {
        id:   p.class_id   ?? null,
        name: p.class_name ?? null,
        slug: p.class_slug ?? null,
      },
    },
  }
}

// =============================================================================
// Paper service functions
// =============================================================================

/**
 * Fetch a single published paper by id with full subject/class detail.
 * Return shape includes nested `subjects` for backward compatibility.
 *
 * @param {number|string} id
 * @returns {Promise<{ data: PaperDetail }>}
 */
export const getPaper = async (id) => {
  const data = await apiFetch(`/api/v1/papers/${id}`)
  return { data: adaptPaperDetail(data) }
}

/**
 * Fetch a paper by slug (e.g. "class-10-maths-annual-2024-42").
 * Extracts the numeric id from the slug suffix and delegates to getPaper().
 *
 * @param {string} slug
 * @returns {Promise<{ data: PaperDetail }>}
 */
export const getPaperBySlug = async (slug) => {
  const parts = slug.split('-')
  const id = parseInt(parts[parts.length - 1], 10)
  if (isNaN(id)) throw new Error('Paper not found')
  return getPaper(id)
}

/**
 * Fetch the N most recently uploaded published papers.
 *
 * @param {number} [limit=10]
 * @returns {Promise<{ data: PaperSummary[] }>}
 */
export const getRecentPapers = async (limit = 10) => {
  const res = await apiFetch(`/api/v1/papers?sort=recent&limit=${limit}`)
  // API returns { data: [...], count, limit }
  return { data: res.data }
}

/**
 * Fetch the N most downloaded published papers.
 *
 * @param {number} [limit=10]
 * @returns {Promise<{ data: PaperSummary[] }>}
 */
export const getPopularPapers = async (limit = 10) => {
  const res = await apiFetch(`/api/v1/papers?sort=popular&limit=${limit}`)
  // API returns { data: [...], count, limit }
  return { data: res.data }
}

/**
 * Returns the list of exam types.
 * Pure data — no API call needed.
 */
export const getExamTypes = () =>
  Promise.resolve({ data: { exam_types: EXAM_TYPES } })

/**
 * Record a paper download by calling the FastAPI download endpoint.
 * POST /api/v1/papers/{id}/download returns 204 No Content on success.
 *
 * @param {number|string} id
 * @returns {Promise<void>}
 */
export const recordDownload = async (id) => {
  await apiFetch(`/api/v1/papers/${id}/download`, { method: 'POST' })
}
